### PR TITLE
Sync contacts from identity store

### PR DESCRIPTION
**Pull request by KaitCrawford**

### CODE REVIEW

**Comment by KaitCrawford** [comment]
@hodgestar, @miltontony first pass at this. please let me know if I'm on the right track

File: casepro/backend/junebug.py [hunk 1]
```diff
@@ -115,7 +175,28 @@ def pull_contacts(
...
+
+        # all identities deleted in the Identity Store in the time window
+        deleted_identities = identity_store.get_identities(
+            optout__optout_type='forget', updated_at__gte=modified_after,
+            updated_at__lte=modified_before)
+
+        # the method expects fetches not lists so I faked it
+        return sync_local_to_changes(
+            org, IdentityStoreContactSyncer(), [identities_to_update],
+            [deleted_identities], progress_callback)
```
Comment (hodgestar) [inline casepro/backend/junebug.py]: I don't think we want the square brackets here -- or at least it seems odd to have a list of iterators.

**Comment by hodgestar** [comment]
The direction looks good. I'm going to look at the test failures quickly.

**Comment by hodgestar** [comment]
```

File: casepro/backend/junebug.py [hunk 2]
```diff
@@ -115,7 +175,28 @@ def pull_contacts(
             tuple of the number of contacts created, updated, deleted and
             ignored
         """
-        return (0, 0, 0, 0)
+        identity_store = self.identity_store
+
+        # all identities created in the Identity Store in the time window
+        new_identities = identity_store.get_identities(
+            created_at__gte=modified_after)
```
Comment (KaitCrawford) [inline casepro/backend/junebug.py]: Wouldn't that mean we would get all existing identities? Since `modified_before` is probably the current time. Don't we instead want all the identities that were created since the last time we ran this?

**Comment by KaitCrawford** [comment]
@hodgestar I haven't implemented fields and a field syncer because I don't actually know what they are or if we want to sync them. Can you help me understand them?

File: casepro/backend/junebug.py (hunk 1 above)
Comment (KaitCrawford) [inline casepro/backend/junebug.py]: The method this is being passed to expects a list of "fetches" where each "fetch" contains a number of contacts. I'm a little uncertain as to what a "fetch" actually is but it seemed like a list of iterators is pretty much exactly what they are expecting.

File: casepro/backend/junebug.py (hunk 2 above)
Comment (hodgestar) [inline casepro/backend/junebug.py]: I meant adding it, rather than replacing it -- so all identities created between `modified_after` and `modified_before`.

Comment (KaitCrawford) [inline casepro/backend/junebug.py]: ... ... ... good point :smile:

File: casepro/backend/junebug.py [hunk 3]
```diff
@@ -40,6 +46,60 @@ def get_addresses(self, uuid):
...
+            i.get('details').get('name') is not "removed"
+        )
+
+
+class IdentityStoreContact(object):
+    """
+    Holds identity data for syncing
+    """
+    def __init__(self, json_data):
+        self.id = json_data.get('id')
```
Comment (hodgestar) [inline casepro/backend/junebug.py]: I wonder whether it would be better to just set all the values from the data as attributes? E.g.
```python
for k, v in json_data.items():
    setattr(self, k, v)
```
?

File: casepro/backend/junebug.py [hunk 4]
```diff
@@ -40,6 +46,60 @@ def get_addresses(self, uuid):
...
+    model = Contact
+    remote_id_attr = 'id'
+
+    def local_kwargs(self, org, remote):
+        return {
+            'org': org,
+            'uuid': remote.id,
+            'name': remote.name,
+            'is_stub': False,
+        }
```
Comment (hodgestar) [inline casepro/backend/junebug.py]: We probably need all of the values that are returned by the `rapidpro` backend. The one's we're missing are:

* `language'`
* `is_blocked`
* `is_stub`
* `fields` (maybe can be an empty list to start with)
* `Contact.SAVE_GROUPS_ATTR` (maybe can be an empty list to start with)
```

File: casepro/backend/junebug.py [hunk 5]
```diff
@@ -40,6 +46,60 @@ def get_addresses(self, uuid):
...
+
+    def update_required(self, local, remote, remote_as_kwargs):
+        if local.is_stub or local.name != remote.name:
+            return True
+
+        return not is_dict_equal(
+            local.get_fields, remote.fields, ignore_none_values=True)
```
Comment (hodgestar) [inline casepro/backend/junebug.py]: This should look almost exactly like the one in the rapidpro backend (except for perhaps a few minor changes in names of fields, etc). Frankly I'm a bit surprised this is needed (because in theory one can implement it entirely generically in terms of `local_kwargs`.

File: casepro/backend/junebug.py [hunk 6]
```diff
@@ -40,6 +46,60 @@ def get_addresses(self, uuid):
...
+
```
Comment (hodgestar) [inline casepro/backend/junebug.py]: We probably also need a  `delete_local()` here. I suspect the implementation should be identical to the `rapidpro` backend one.

File: casepro/backend/junebug.py (hunk 4 above)
Comment (KaitCrawford) [inline casepro/backend/junebug.py]: We don't currently store language in the Identity Store. What should it default to?

Comment (hodgestar) [inline casepro/backend/junebug.py]: @imsickofmaps Is there any way to get the users language currently? (apologies if @KaitCrawford has already asked you)

Comment (imsickofmaps) [inline casepro/backend/junebug.py]: So far we've been using the `preferred_langage` field in `details`. See: https://github.com/praekelt/seed-services-cli and demo: https://github.com/praekelt/seed-services-cli/blob/develop/demo_import_identities.csv

Comment (KaitCrawford) [inline casepro/backend/junebug.py]: @hodgestar @imsickofmaps I'm not really sure how to handle the "`is_blocked` (this probably is similar to opt outs)" section. Opt-outs are a related model that aren't returned when querying the Identity Store. I could get the opt-outs for each identity after we've gotten them but the opt-out API endpoint doesn't allow retrieve actions (whoops?). Should I add that or is there some other way we could do this?

Comment (hodgestar) [inline casepro/backend/junebug.py]: @KaitCrawford Let's just set it to `False` for and an leave a TODO -- we can address it in a separate PR (and possibly add the needed methods to the identity API).

File: casepro/backend/junebug.py [hunk 7]
```diff
@@ -40,6 +46,60 @@ def get_addresses(self, uuid):
         return (
             a['address'] for a in addresses if a.get('address') is not None)
 
+    def get_identities(self, **kwargs):
+        '''Get the list of identities filtered by the given kwargs.'''
+        url = '%s/api/v1/identities/search/?' % (self.base_url)
+        for key, value in kwargs.iteritems():
+            url = '%s%s=%s&' % (url, key, value)
```
Comment (rudigiesler) [inline casepro/backend/junebug.py]: Maybe we should use the `params` dictionary below instead of manually doing it ourselves here? That should take care of properly encoding the params (if there are any special characters). Also, I don't think we need the `default=True` parameter here, I think that only applies to the `/addresses/` endpoint.

File: casepro/backend/junebug.py [hunk 8]
```diff
@@ -40,6 +46,74 @@ def get_addresses(self, uuid):
         return (
             a['address'] for a in addresses if a.get('address') is not None)
 
+    def get_identities(self, **kwargs):
+        '''Get the list of identities filtered by the given kwargs.'''
+        url = '%s/api/v1/identities/search/?' % (self.base_url)
+        params = {}
+        for key, value in kwargs.iteritems():
+            params[key] = value
```
Comment (hodgestar) [inline casepro/backend/junebug.py]: You can delete the three lines above and just put `**params` in the function signature.

File: casepro/backend/junebug.py [hunk 9]
```diff
@@ -40,6 +46,74 @@ def get_addresses(self, uuid):
         return (
             a['address'] for a in addresses if a.get('address') is not None)
 
+    def get_identities(self, **kwargs):
+        '''Get the list of identities filtered by the given kwargs.'''
+        url = '%s/api/v1/identities/search/?' % (self.base_url)
```
Comment (hodgestar) [inline casepro/backend/junebug.py]: Just checking whether you intended to write `(self.base_url,)` here instead of `(self.base_url)`?

File: casepro/backend/junebug.py [hunk 10]
```diff
@@ -40,6 +46,74 @@ def get_addresses(self, uuid):
...
+        params = {}
+        for key, value in kwargs.iteritems():
+            params[key] = value
+
+        identities = self.get_paginated_response(
+            url, params=params)
+
+        return (
+            IdentityStoreContact(i) for i in identities if
+            i.get('details').get('name') is not "removed"
```
Comment (hodgestar) [inline casepro/backend/junebug.py]: This is perhaps worth a comment about optouts, since otherwise people will wonder what this `name is not "removed"` bit is all about.

File: casepro/backend/junebug.py [hunk 11]
```diff
@@ -40,6 +46,74 @@ def get_addresses(self, uuid):
...
+    """
+    Holds identity data for syncing
+    """
+    def __init__(self, json_data):
+        for k, v in json_data.items():
+            setattr(self, k, v)
+
+        # Languages in the idenity store have the country code at the end
+        remote_language = json_data.get('details').get('preferred_langage')
+        language = remote_language[:remote_language.index('_')]
```
Comment (hodgestar) [inline casepro/backend/junebug.py]: It's probably safer to do `_, _, language = remote_language.rpartition('_')`. It's a bit arcane but guards against edge cases where the string doesn't contain an `_`.

File: casepro/backend/tests/test_junebug.py [hunk 12]
```diff
@@ -13,20 +13,197 @@ def setUp(self):
...
+            self.unicef, '2016-03-14T10:25:00', '2016-03-14T10:21:00')
+        self.assertEqual(created, 1)
+        self.assertEqual(updated, 0)
+        self.assertEqual(deleted, 0)
+        self.assertEqual(ignored, 0)
+        self.assertEqual(Contact.objects.count(), 1)
+
+    @responses.activate
+    def test_pull_contacts_recently_updated(self):
+        Contact.objects.all().delete()
```
Comment (hodgestar) [inline casepro/backend/tests/test_junebug.py]: Are `Contact` objects not cleaned up in between each test automatically? Or are there some fixtures that are automatically loaded that we don't want?

File: casepro/backend/tests/test_junebug.py [hunk 13]
```diff
@@ -402,6 +579,55 @@ def request_callback(request):
         self.assertEqual(sorted(res), sorted(['+1234', '+4321']))
 
     @responses.activate
+    def test_get_identities(self):
+        '''The get_identities function should call the correct URL, and return
+        the relevant identities.'''
+        identity_store = IdentityStore(
+            'http://identitystore.org/', 'auth-token', 'msisdn')
+
+        def request_callback(request):
```
Comment (hodgestar) [inline casepro/backend/tests/test_junebug.py]: Could we put this callback in a separate method like you did with the others? That looked really clean.

**Comment by hodgestar** [comment]
I left some minor comments.

One bigger comment is that should we check that the contacts created by `pull_contacts` in the `pull_contacts` tests are what we expect them to be?

**Comment by KaitCrawford** [comment]
@hodgestar @rudigiesler ready for re-review in the morning

File: casepro/backend/junebug.py [hunk 14]
```diff
@@ -40,6 +46,73 @@ def get_addresses(self, uuid):
...
+class IdentityStoreContact(object):
+    """
+    Holds identity data for syncing
+    """
+    def __init__(self, json_data):
+        for k, v in json_data.items():
+            setattr(self, k, v)
+
+        # Languages in the idenity store have the country code at the end
+        remote_language = json_data.get('details').get('preferred_langage')
```
Comment (hodgestar) [inline casepro/backend/junebug.py]: Is the spelling of `preferred_langage` correct? It seems to be missing a `u` here and in tests?

**Comment by hodgestar** [comment]
One last query about the spelling of `preferred_langage`, otherwise looks good!

**Comment by KaitCrawford** [comment]
@hodgestar ready again :)

**Comment by hodgestar** [comment]
:+1:

File: casepro/backend/junebug.py [hunk 15]
```diff
@@ -115,7 +188,28 @@ def pull_contacts(
             tuple of the number of contacts created, updated, deleted and
             ignored
         """
-        return (0, 0, 0, 0)
+        identity_store = self.identity_store
+
+        # all identities created in the Identity Store in the time window
+        new_identities = identity_store.get_identities(
+            created_at__gte=modified_after, created_at__lte=modified_before)
```
Comment (rudigiesler) [inline casepro/backend/junebug.py]: The issue I see here is that I don't think that the identity store supports multiple arguments on its `search` endpoint. That should probably be fixed in the identity store, so we should probably create an issue there (also, this sync probably won't work properly since currently the identity store will only filter on the first argument).

Comment (KaitCrawford) [inline casepro/backend/junebug.py]: @rudigiesler I've just tested searching on multiple arguments locally and it works fine :)

Comment (rudigiesler) [inline casepro/backend/junebug.py]: Oh that's interesting. Yes an AND is how the django filter arguments work, and it's what we want. Looking at this though, I thought that multiple arguments wouldn't work: https://github.com/praekelt/seed-identity-store/blob/develop/identities/views.py#L63

Comment (KaitCrawford) [inline casepro/backend/junebug.py]: You're right. I have no idea why that works then :confused:

**Comment by miltontony** [comment]
@imsickofmaps can we get a review here please?
Just making sure you're on the same page on what's been done here

File: casepro/backend/junebug.py [hunk 16]
```diff
@@ -40,6 +46,73 @@ def get_addresses(self, uuid):
...
+        url = '%s/api/v1/identities/search/?' % self.base_url
+
+        identities = self.get_paginated_response(
+            url, params=params)
+
+        # Users who opt to be forgotten from the system have their details
+        # stored as 'removed'
+        return (
+            IdentityStoreContact(i) for i in identities if
+            i.get('details').get('name') is not "removed"
```
Comment (imsickofmaps) [inline casepro/backend/junebug.py]: This isn't quite right. See https://github.com/praekelt/seed-identity-store/blob/develop/identities/models.py#L76-L86

If someone's requested forget me, then when you search for them by address_type/address you won't find them. If you happen to access them directly using `/identity/uuid/`you'll see their `details["addresses"] == {}`. Any optional details ("name" is optional) would be `redacted` not `removed`.

File: casepro/backend/junebug.py [hunk 17]
```diff
@@ -40,6 +46,73 @@ def get_addresses(self, uuid):
...
+
+class IdentityStoreContact(object):
+    """
+    Holds identity data for syncing
+    """
+    def __init__(self, json_data):
+        for k, v in json_data.items():
+            setattr(self, k, v)
+
+        # Languages in the idenity store have the country code at the end
```
Comment (imsickofmaps) [inline casepro/backend/junebug.py]: minor typo `identity`

File: casepro/backend/junebug.py (hunk 16 above)
Comment (KaitCrawford) [inline casepro/backend/junebug.py]: I don't think this integration can currently handle the "'name' is optional" side note there. But I'll change 'removed' to 'redacted' quickly.

File: casepro/backend/junebug.py (hunk 15 above)
Comment (imsickofmaps) [inline casepro/backend/junebug.py]: The search endpoint isn't really for this. I would use the `/identities/` base and update https://github.com/praekelt/seed-identity-store/blob/develop/identities/views.py#L37 to have more of the fields you want to filter on

File: casepro/backend/junebug.py (hunk 16 above)
Comment (imsickofmaps) [inline casepro/backend/junebug.py]: I'm unsure how RapidPro would know the "name" of millions of randoms that SMS in, but we can ask them

File: casepro/backend/junebug.py (hunk 15 above)
Comment (KaitCrawford) [inline casepro/backend/junebug.py]: @imsickofmaps I'm inclined to make this a TODO or github issue to be tackled later. The search method is currently operational whereas the filters on the identity store need some work. Currently one can only filter on details but we need to be able to filter on a date range which requires a FilterSet to be implemented.

File: casepro/backend/junebug.py (hunk 16 above)
Comment (KaitCrawford) [inline casepro/backend/junebug.py]: I've just realised a great big problem with this anyway. Because of this filter we never actually return any identities that have been deleted :facepalm:

**Comment by KaitCrawford** [comment]
@imsickofmaps, @hodgestar @rudigiesler @miltontony ready for re-review for whoever wants to.

File: casepro/backend/junebug.py [hunk 18]
```diff
@@ -40,6 +46,75 @@ def get_addresses(self, uuid):
...
+    """
+    Holds identity data for syncing
+    """
+    def __init__(self, json_data):
+        for k, v in json_data.items():
+            setattr(self, k, v)
+
+        # Languages in the identity store have the country code at the end
+        remote_language = json_data.get('details').get('preferred_language')
+        language, _, _ = remote_language.partition('_')
```
Comment (rudigiesler) [inline casepro/backend/junebug.py]: We can just do `self.language` here to avoid having to set it below (since I don't think that we are using `language` anywhere else).

Also, what if there is no `preferred_language` in `details`? That would mean that we are doing a `None.partition` here, which would result in an error. Maybe this should be in an `if remote_language is not None:`, that way no language is set if the preferred language is not set.

As far as I can see from the identity store, only `addresses` and `default_address_type` are required in the `details` field.

File: casepro/backend/junebug.py [hunk 19]
```diff
@@ -115,7 +190,28 @@ def pull_contacts(
...
+        # all identities modified in the Identity Store in the time window
+        modified_identities = identity_store.get_identities(
+            updated_from=modified_after,
+            updated_to=modified_before)
+
+        identities_to_update = list(chain(modified_identities, new_identities))
+
+        # all identities deleted in the Identity Store in the time window
+        deleted_identities = list(identity_store.get_identities(
+            optout_type='forget', updated_from=modified_after,
```
Comment (rudigiesler) [inline casepro/backend/junebug.py]: Should this not be `optout_set__optout_type`? Since it's a foreign key to the OptOut object, which has the `optout_type` field.

Comment (rudigiesler) [inline casepro/backend/junebug.py]: Ah, I see that this is a specified filter, not a django query. Ignore that comment.

**Comment by rudigiesler** [comment]
Just that one comment from me, looking good otherwise. 🍰  for this long PR

File: casepro/backend/junebug.py (hunk 15 above)
Comment (KaitCrawford) [inline casepro/backend/junebug.py]: This is waiting for a decision on https://github.com/praekelt/seed-identity-store/pull/24

Comment (KaitCrawford) [inline casepro/backend/junebug.py]: @imsickofmaps how would you like to proceed here?

**Comment by hodgestar** [comment]
I'm :+1: on landing this and then opening a new PR if the approach used in the identity store changes from what @KaitCrawford has proposed in praekelt/seed-identity-store#24.
